frameworks/keyed/vanillajs-2/src/Main.js

```diff
@@ -0,0 +1,205 @@
+                //console.log("add");
+                this.add();
+            }
+            else if (e.target.matches('#run')) {
+                e.stopPropagation();
+                //console.log("run");
+                this.run();
+            }
+            else if (e.target.matches('#update')) {
+                e.stopPropagation();
+                //console.log("update");
+                this.update();
+            }
+            else if (e.target.matches('#runlots')) {
+                e.stopPropagation();
+                //console.log("runLots");
+                this.runLots();
+            }
+            else if (e.target.matches('#clear')) {
+                e.stopPropagation();
+                //console.log("clear");
+                this.clear();
+            }
+            else if (e.target.matches('#swaprows')) {
+                e.stopPropagation();
+                //console.log("swapRows");
+                this.swapRows();
+            }
+            else if (e.target.matches('.remove')) {
+                let id = getParentId(e.target);
+                let idx = this.data.findIndex(row => row.id === id);
+                //console.log("delete",idx);
+                this.delete(idx);
+            }
+            else if (e.target.matches('.lbl')) {
+                let id = getParentId(e.target);
+                let idx = this.data.findIndex(row => row.id === id);
+                //console.log("select",idx);
+                this.select(idx);
+            }
+        });
+        this.tbody = document.getElementById("tbody");
+    }
+    run() {
+        this.removeAllRows();
+        this.data = [];
+        this.appendRows(buildData(1000));
+        this.unselect();
+    }
+    add() {
+        this.appendRows(buildData(1000));
+    }
+    update() {
+        for (let i=0;i<this.data.length;i+=10) {
+            this.data[i].label += ' !!!'
+            this.tbody.childNodes[i].firstChild.nextSibling.firstChild.firstChild.data = this.data[i].label;
+        }
+    }
+    unselect() {
+        if (this.selectedRow !== undefined) {
+            this.selectedRow.className = "";
+            this.selectedRow = undefined;
+        }
+    }
+    select(idx) {
+        this.unselect();
+        this.selectedRow = this.tbody.childNodes[idx];
+        this.selectedRow.className = "danger";
+    }
+    delete(idx) {
+        // Remove that row from the DOM
+        this.tbody.childNodes[idx].remove();
+        this.data.splice(idx, 1);
+    }
+    removeAllRows() {
+        // ~258 msecs
+        // for(let i=this.rows.length-1;i>=0;i--) {
+        //     tbody.removeChild(this.rows[i]);
+        // }
+        // ~251 msecs
+        // for(let i=0;i<this.rows.length;i++) {
+        //     tbody.removeChild(this.rows[i]);
+        // }
+        // ~216 msecs
+        // var cNode = tbody.cloneNode(false);
+        // tbody.parentNode.replaceChild(cNode ,tbody);
+        // ~212 msecs
+        this.tbody.textContent = "";
```

Comment:
You could also try creating a new `tbody` element and replacing the existing `tbody` with it

Reply:
"Replacing the tbody with an empty one is O(1) in Chrome, whereas textContent = '' is O(n) looking at the Chrome code" - so, it turns out it's better not to do it?

Reply:
Yeah you want to replace the node instead of clearing text content

Reply:
Well, yes. If you also take into account, for example, that logically there can be a bunch of event listeners on tbody, then it’s better not to do this as an example to other people. So, in principle, in this particular speed task, this is not critical. But, it’s really better to leave textContent then.

Reply:
The event listeners can be on the table.

Reply:
@syduki cloneNode() does not copy event listeners. At the very least, this solution cannot be used for vanilla-js. But, in general, in a specific task of cleaning a table, it is not some kind of direct cheating if you change the tbody.

Reply:
It's arguable a cheat too, as if this was a stateful DOM element like a custom element, then replacing it with an empty clone is also not acceptable and in the real world would cause major problems. I'm not sure if we should push ivi to change its implementation here (or add a note) or instead allow this. That's up to @krausest to decide.